core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -567,28 +567,35 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
         return result;
     }
 
+    public Engine.NoOp prepareNoOpOnReplica(String type, String id, long seqNo, String reason) {
+        verifyReplicationTarget();
+        final Term uid = extractUid(type, id);
+        long startTime = System.nanoTime();
+        return new Engine.NoOp(uid, seqNo, primaryTerm, Engine.Operation.Origin.REPLICA, startTime, reason);
+    }
+
+    public Engine.NoOpResult noOp(Engine.NoOp noOp) throws IOException {
```

Comment:
I think we should call these something like markSeqNoAsNoOp. It feels weird to the the shard to do nothing ;) The previous method can maybe called preparingMarkingSeqNoAsNoOp?